server/app/services/scheduler/filter/affinity.rb

```diff
@@ -89,34 +69,65 @@ def split_affinity(affinity)
         end
       end
 
-      # @param [String] comparator
+      # @param flags [String]
       # @return [Boolean]
       def soft?(flags)
         flags && flags.include?('~')
       end
 
-      # @param [HostNode] node
-      # @param [String] value
+      # @param node [HostNode]
+      # @param value [String]
+      # @return [Boolean]
       def node_match?(node, value)
-        node.name == value
+        value_matches?(node.name, value)
       end
 
-      # @param [HostNode] node
-      # @param [String] value
+      # @param node [HostNode]
+      # @param value [String]
+      # @return [Boolean]
       def container_match?(node, value)
         node.containers.where(name: value).exists?
+        container_names = node.containers.map { |c|
+          c.name
+        }
+        container_names.any?{ |n| value_matches?(n, value) }
       end
 
-      # @param [HostNode] node
-      # @param [GridService, nil] value
-      def service_match?(node, value)
-        value && node.grid_service_instances.where(grid_service: value).exists?
+      # @param node [HostNode]
+      # @param value [String]
+      # @param service [GridService,NilClass]
+      # @return [Boolean]
+      def service_match?(node, value, service)
+        match_with_stack = regex?(value) ? value[1...-1].include?('\/') : value.include?('/')
+        value = "#{service.stack.name}/#{value}" unless match_with_stack
+        service_names = node.grid_service_instances.includes(:grid_service).map { |i|
+          "#{i.grid_service.stack.name}/#{i.grid_service.name}"
```

Comment:
For some reason, I seem to have `GridServiceInstance` objects that are missing their `grid_service`... might just be a development fluke because they're very old, been hanging around in my development mongo DB since early 2017... but this is first time that they've exploded :o

```
E, [2018-01-05T10:17:56.387558 #13] ERROR -- GridServiceDeployer: Unknown error (development/service-affinity/web): NoMethodError undefined method `stack' for nil:NilClass
E, [2018-01-05T10:17:56.387614 #13] ERROR -- GridServiceDeployer: /app/app/services/scheduler/filter/affinity.rb:104:in `block in service_match?'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:678:in `yield_document'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:130:in `block in each'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:129:in `each'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:129:in `each'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:309:in `map'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:309:in `map'
/usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual.rb:20:in `map'
/app/app/services/scheduler/filter/affinity.rb:103:in `service_match?'
/app/app/services/scheduler/filter/affinity.rb:42:in `match_affinity?'
/app/app/services/scheduler/filter/affinity.rb:17:in `block (2 levels) in for_service'
/app/app/services/scheduler/filter/affinity.rb:16:in `select'
/app/app/services/scheduler/filter/affinity.rb:16:in `block in for_service'
/app/app/services/scheduler/filter/affinity.rb:12:in `each'
/app/app/services/scheduler/filter/affinity.rb:12:in `for_service'
/app/app/services/grid_service_scheduler.rb:73:in `block in filter_nodes'
/app/app/services/grid_service_scheduler.rb:72:in `each'
/app/app/services/grid_service_scheduler.rb:72:in `filter_nodes'
/app/app/services/grid_service_scheduler.rb:52:in `block in select_node'
/app/app/services/grid_service_scheduler.rb:51:in `synchronize'
/app/app/services/grid_service_scheduler.rb:51:in `select_node'
/app/app/services/grid_service_deployer.rb:136:in `block in instance_count'
/app/app/services/grid_service_deployer.rb:134:in `times'
/app/app/services/grid_service_deployer.rb:134:in `instance_count'
/app/app/services/grid_service_deployer.rb:52:in `deploy'
/app/app/jobs/grid_service_scheduler_worker.rb:68:in `deploy'
/app/app/jobs/grid_service_scheduler_worker.rb:13:in `block in watch'
/app/app/jobs/grid_service_scheduler_worker.rb:11:in `loop'
/app/app/jobs/grid_service_scheduler_worker.rb:11:in `watch'
```